ci/generate_combinations.py

```diff
@@ -0,0 +1,398 @@
+        )
+        mods_missing_from_actual = expect_modules - actual_modules
+        mods_missing_from_expect = actual_modules - expect_modules
+        if mods_missing_from_actual:
+            logger.info(
+                f"{actual_name}['modules'] for Qt {version} is missing {mods_missing_from_actual}"
+            )
+        if mods_missing_from_expect:
+            logger.info(
+                f"{expect_name}['modules'] for Qt {version} is missing {mods_missing_from_expect}"
+            )
+        return bool(mods_missing_from_actual) or bool(mods_missing_from_expect)
+
+    def to_set(a_list: Union[List[str], List[Dict]]) -> Set:
+        if len(a_list) == 0:
+            return set()
+        if isinstance(a_list[0], str):
+            return set(a_list)
+        assert isinstance(a_list[0], Dict)
+        return set([str(a_dict) for a_dict in a_list])
+
+    def report_difference(
+        superset: Set, subset: Set, subset_name: str, key: str
+    ) -> bool:
+        """Return True if difference detected. Print description of difference."""
+        missing_from_superset = sorted(superset - subset)
+        if not missing_from_superset:
+            return False
+        logger.info(f"{subset_name}['{key}'] is missing these entries:")
+        if key in list_of_str_keys:
+            logger.info(format(missing_from_superset))
+            return True
+        for el in missing_from_superset:
+            logger.info(format(el))
+        return True
+
+    for root_key in actual_combos.keys():
+        if root_key in skipped_keys:
+            continue
+
+        logger.info(f"\nComparing {root_key}:\n{'-' * 40}")
+        if root_key == "modules":
+            for actual_row, expect_row in zip(
+                actual_combos[root_key], expected_combos[root_key]
+            ):
+                assert actual_row["qt_version"] == expect_row["qt_version"]
+                has_difference |= compare_modules_entry(actual_row, expect_row)
+            continue
+
+        actual_set = to_set(actual_combos[root_key])
+        expected_set = to_set(expected_combos[root_key])
+        has_difference |= report_difference(
+            expected_set, actual_set, actual_name, root_key
+        )
+        has_difference |= report_difference(
+            actual_set, expected_set, expect_name, root_key
+        )
+
+    return has_difference
+
+
+def alphabetize_modules(combos: Dict[str, Union[List[Dict], List[str]]]):
+    for i, item in enumerate(combos["modules"]):
+        combos["modules"][i]["modules"] = sorted(item["modules"])
+
+
+def write_combinations_json(
+    combos: Dict[str, Union[List[Dict], List[str]]],
+    filename: Path,
+    is_use_pretty_print: bool = True,
+):
+    logger.info(f"Write file {filename}")
+    json_text = (
+        pretty_print_combos(combos)
+        if is_use_pretty_print
+        else json.dumps([combos], sort_keys=True, indent=2)
+    )
+    if filename.write_text(json_text, encoding="utf_8") == 0:
```

Comment:
By default, this script writes the json file using my `pretty_print_combos` function, which is not well tested. It's possible that this code could write the wrong thing in the json file.

The `json.dumps` function is well tested, and I think it would work do a better job at printing the json file. However, with the parameter `indent=2`, it writes 870 lines of json data, which is not so easy to read. If you turn off `indent`, it just writes the whole object on a single line.

Reply:
When the action involved, no one doesn't read json directly. Otherwise, we will use the utility.
It is ok to use `json.dump` with indent.  Github help us to see difference in PR diff.

Reply:
https://pypi.org/project/jsoncomparison/ may be usable to show diff.

Reply:
> It is ok to use `json.dump` with indent. Github help us to see difference in PR diff.

Sounds good to me.

> https://pypi.org/project/jsoncomparison/ may be usable to show diff.

I tried using `jsoncomparison` to show the diff for this program's output, but I am not happy with the result. Here's a link to some recent output: https://gist.github.com/ddalcino/687515d367d543372fe5924e6f0c5372

I think `jsoncomparison` would be perfectly fine for checking whether a difference exists or not, and thereby deciding whether to open a PR. I think that the output shown in gist above will be useless to us, so I just won't print it. Github will be a lot better at showing us the diff anyway.